File: mistral-ocr-web/src/lib/markdown-generator.ts
import { OCRResponse, FigureCounter } from './types';

/**
 * 複数のOCRレスポンスから統合されたMarkdownを生成する
 *
 * @param ocrResponses - OCRレスポンスの配列
 * @param includePageHeaders - ページヘッダー（# Page N）を含めるかどうか（デフォルト: false）
 * @returns 統合されたMarkdown文字列
 */
export function generateMarkdown(ocrResponses: OCRResponse[], includePageHeaders: boolean = false): string {
  const counter: FigureCounter = { imageCount: 0, tableCount: 0 };
  let markdown = '';
  let globalPageNumber = 0;

  for (const response of ocrResponses) {
    for (const page of response.pages) {
      globalPageNumber++;

      // ページヘッダーをオプションで追加
      if (includePageHeaders) {
        markdown += `# Page ${globalPageNumber}\n\n`;
      }

      let pageMarkdown = page.markdown || '';

      // 画像の処理と図番号の付与
      if (page.images && page.images.length > 0) {
        for (const image of page.images) {
          counter.imageCount++;

          // ![img-X.jpeg](img-X.jpeg) を Base64画像に置換
          // 正規表現で画像参照を検出
          const imgPattern = new RegExp(
            `!\\[${escapeRegExp(image.id)}\\]\\(${escapeRegExp(image.id)}\\)`,
            'g'
          );

          // Base64画像への置換文字列を作成
          const replacement = `![図${counter.imageCount}](${image.image_base64})\n\n**図${counter.imageCount}**`;

          pageMarkdown = pageMarkdown.replace(imgPattern, replacement);
        }
      }

      // 表の処理と表番号の付与（将来的な拡張用）
      // Mistral APIのtables配列に情報がある場合の処理
      if (page.tables && page.tables.length > 0) {
        // テーブルの処理ロジック
        // 現在はtables配列が空なので、将来の拡張として保持
        for (const table of page.tables) {
          counter.tableCount++;
          // 表の前後に番号を挿入する処理
          // （具体的な実装はMistral APIのtable構造に依存）
        }
      }

      // Markdown内の表構文（|で区切られた行）を検出して表番号を付与
      pageMarkdown = addTableNumbers(pageMarkdown, counter);

      markdown += pageMarkdown + '\n\n---\n\n';
    }
  }

  return markdown.trim();
}

/**
 * Markdown内の表構文を検出して表番号を付与する
 *
 * @param markdown - 処理対象のMarkdown文字列
 * @param counter - 図表カウンター
 * @returns 表番号が付与されたMarkdown文字列
 */
function addTableNumbers(markdown: string, counter: FigureCounter): string {
  // Markdown表の開始パターン（| で始まる行）
  const tablePattern = /^(\|.+\|)\s*$/gm;
  const lines = markdown.split('\n');
  const result: string[] = [];
  let inTable = false;
  let tableStartIndex = -1;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    const isTableLine = /^\|.+\|/.test(line.trim());

    if (isTableLine && !inTable) {
      // 表の開始
      inTable = true;
      tableStartIndex = i;
      counter.tableCount++;
      result.push(`**表${counter.tableCount}**\n`);
      result.push(line);
    } else if (isTableLine && inTable) {
      // 表の継続
      result.push(line);
    } else if (!isTableLine && inTable) {
      // 表の終了
      inTable = false;
      result.push(line);
    } else {
      // 通常の行
      result.push(line);
    }
  }

  return result.join('\n');
}

/**
 * 正規表現で使用する特殊文字をエスケープする
 *
 * @param string - エスケープする文字列
 * @returns エスケープされた文字列
 */
function escapeRegExp(string: string): string {
  return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

/**
 * Markdownファイルのダウンロード用のファイル名を生成する
 *
 * @returns ファイル名（例: ocr_result_20260101_143000.md）
 */
export function generateFilename(): string {
  const now = new Date();
  const year = now.getFullYear();
  const month = String(now.getMonth() + 1).padStart(2, '0');
  const day = String(now.getDate()).padStart(2, '0');
  const hours = String(now.getHours()).padStart(2, '0');
  const minutes = String(now.getMinutes()).padStart(2, '0');
  const seconds = String(now.getSeconds()).padStart(2, '0');

  return `ocr_result_${year}${month}${day}_${hours}${minutes}${seconds}.md`;
}

File: src/mistral_ocr/markdown_generator.py
"""
Mistral OCR APIのレスポンスからMarkdownを生成するモジュール。
TypeScript版の図表自動付番ロジックをPythonに移植。
"""

import base64
from datetime import datetime
from pathlib import Path
import re
from typing import Any

from .types import FigureCounter


def _save_image(
    image_base64: str,
    output_dir: Path,
    figure_number: int
) -> str:
    """
    Base64画像をファイルとして保存し、相対パスを返す。

    Args:
        image_base64: Base64エンコードされた画像データ (Data URI形式も可)
        output_dir: 出力先ディレクトリ
        figure_number: 図番号

    Returns:
        Markdownで使用する相対パス (例: images/figure_001.jpg)
    """
    # imagesディレクトリを作成
    images_dir = output_dir / "images"
    images_dir.mkdir(exist_ok=True)

    # Data URI形式の場合、Base64部分のみを抽出
    if "," in image_base64:
        base64_data = image_base64.split(",", 1)[1]
    else:
        base64_data = image_base64

    # Base64デコードしてファイルに保存
    image_data = base64.b64decode(base64_data)
    filename = f"figure_{figure_number:03d}.jpg"
    filepath = images_dir / filename
    filepath.write_bytes(image_data)

    # 相対パスを返す
    return f"images/{filename}"


def generate_markdown(
    ocr_response: Any,
    output_dir: Path,
    include_page_headers: bool = False
) -> str:
    """
    OCRレスポンスから統合Markdownを生成。

    Args:
        ocr_response: Mistral OCR APIのレスポンスオブジェクト
        output_dir: 出力先ディレクトリ (画像ファイル保存用)
        include_page_headers: ページヘッダー (# Page N) を含めるか

    Returns:
        統合されたMarkdown文字列
    """
    counter = FigureCounter()
    markdown = ""
    global_page_number = 0

    for page in ocr_response.pages:
        global_page_number += 1

        if include_page_headers:
            markdown += f"# Page {global_page_number}\n\n"

        page_markdown = page.markdown or ""

        # 画像処理と図番号付与
        if page.images:
            for image in page.images:
                counter.image_count += 1

                # Base64画像を外部ファイルとして保存
                image_path = _save_image(
                    image.image_base64,
                    output_dir,
                    counter.image_count
                )

                # 正規表現でパターンマッチ
                pattern = re.compile(
                    rf"!\[{re.escape(image.id)}\]\({re.escape(image.id)}\)",
                    re.MULTILINE
                )

                # 相対パスを使用したMarkdownに置換
                replacement = (
                    f"![図{counter.image_count}]({image_path})\n\n"
                    f"**図{counter.image_count}**"
                )

                page_markdown = pattern.sub(replacement, page_markdown)

        # 表番号付与
        page_markdown = _add_table_numbers(page_markdown, counter)

        markdown += page_markdown + "\n\n---\n\n"

    return markdown.strip()


def _add_table_numbers(markdown: str, counter: FigureCounter) -> str:
    """
    Markdown内の表構文を検出して表番号を付与。

    Args:
        markdown: 処理対象のMarkdown文字列
        counter: 図表カウンター

    Returns:
        表番号が付与されたMarkdown文字列
    """
    lines = markdown.split("\n")
    result = []
    in_table = False

    for line in lines:
        is_table_line = bool(re.match(r"^\s*\|.+\|\s*$", line))

        if is_table_line and not in_table:
            # 表の開始
            in_table = True
            counter.table_count += 1
            result.append(f"**表{counter.table_count}**")
            result.append(line)
        elif is_table_line and in_table:
            # 表の継続
            result.append(line)
        elif not is_table_line and in_table:
            # 表の終了
            in_table = False
            result.append(line)
        else:
            # 通常の行
            result.append(line)

    return "\n".join(result)


def generate_filename() -> str:
    """
    Markdownファイル名を生成 (例: ocr_result_20260113_143000.md)
    """
    now = datetime.now()
    return now.strftime("ocr_result_%Y%m%d_%H%M%S.md")
